File: adderall.c
/* Adderall: A BrainFuck Interpreter */ 

/* 
Command: 
- `>`: increments the data pointer by one (moves the data pointer to the right) 
- `<`: decrements the data pointer by one (moves the data pointer to the left)
- `+`: increments the byte at the data pointer by one 
- `-`: decrements the byte at the data pointer by one 
- `.`: outputs the byte at the data pointer 
- `,`: accepts one byte as input, and stores it in the byte at the data pointer 
- `[`: if the byte at the data pointer is zero, jump it to the next command after `]` command 
- `]`: if the byte at the data pointer is nonzero, jump it back to the command after `[` command
*/

#include <stdio.h>

#define MEM_SIZE 30000                               // max memory size for the program
#define SRC_SIZE  30000                              // max source file size 

char memory[MEM_SIZE];                               // program memory
char* ptr = memory;                                  // data pointer 

char source[SRC_SIZE];                               // source file buffer 
char* ip = source;                                   // instruction pointer 

int loop;                                            // loop counter

void execute() {
    while (*ip) {
        switch (*ip) {
            case '>': ++ptr; break;                  // increments the memory pointer 
            case '<': --ptr; break;                  // decrements the memory pointer 
            case '+': ++*ptr; break;                 // increments the memory byte
            case '-': --*ptr; break;                 // decrements the memory byte 
            case '.': putc(*ptr, stdout); break;     // output the char 
            case ',': *ptr = getc(stdin); break;     // get char as input 

            case '[': {                              // loop start 
                if (!*ptr) {                         // if cell at the pointer is zero    
                    loop = 1;                        // set loop counter to 1 
                    while (loop) {                   
                        ip++;                        // increment instruction pointer 
                        if (*ip == '[') loop++;      // increment loop counter if `[` cmd 
                        if (*ip == ']') loop--;      // decrement loop counter if `]` cmd
                    }
                }
                break;
            }
            case ']': {                              // loop end 
                loop = 1;                            // set loop counter to 1
                while (loop) {
                    ip--;                            // decrement instruction pointer 
                    if (*ip == '[') loop--;          // decrement loop counter if `[` cmd 
                    if (*ip == ']') loop++;          // increment loop counter if `]` cmd 
                }
                ip--;                                // decrement instruction pointer 
                break;
            }
        }
        ip++;                                        // increment instruction pointer 
    }
}

int main(int argc, char* argv[]) {
    if (argc == 1) {
        printf("\n **** Adderall: A Brainfuck Interpreter ****\n\n");
        printf("  1. Type or paste in brainfuck source code\n");
        printf("  2. Use Ctrl-D to run the code\n");
        printf("  3. Use ./adderall [filename] to execute source file\n"); 
        
        fread(source, 1, SRC_SIZE - 1, stdin);       // read source code from stdin 
        execute();                                   // execute the program 
        printf("\n");
    } else if (argc == 2) {
        FILE *file = fopen(argv[1], "r");            // open the file stream 
        fread(source, 1, SRC_SIZE, file);            // read source from file 
        fclose(file);                                // close the file 
        execute();                                   // execute the program 
    } 
    return 0;
}

File: adderall.rs
use std::env;
use std::fs::File;
use std::io::{self, Read};

const MEM_SIZE: usize = 30000;
const SRC_SIZE: usize = 30000;

struct Interpreter {
    memory: [u8; MEM_SIZE],
    ptr: usize,
    source: Vec<u8>,
    ip: usize,
}

impl Interpreter {
    fn new() -> Self {
        Interpreter {
            memory: [0; MEM_SIZE],
            ptr: 0,
            source: Vec::with_capacity(SRC_SIZE),
            ip: 0,
        }
    }

    fn execute(&mut self) {
        while self.ip < self.source.len() {
            match self.source[self.ip] as char {
                '>' => self.ptr = (self.ptr + 1) % MEM_SIZE,
                '<' => self.ptr = (self.ptr + MEM_SIZE - 1) % MEM_SIZE,
                '+' => self.memory[self.ptr] = self.memory[self.ptr].wrapping_add(1),
                '-' => self.memory[self.ptr] = self.memory[self.ptr].wrapping_sub(1),
                '.' => print!("{}", self.memory[self.ptr] as char),
                ',' => {
                    let mut input = [0; 1];
                    if io::stdin().read_exact(&mut input).is_ok() {
                        self.memory[self.ptr] = input[0];
                    }
                }
                '[' => {
                    if self.memory[self.ptr] == 0 {
                        let mut loop_count = 1;
                        while loop_count > 0 {
                            self.ip += 1;
                            match self.source[self.ip] as char {
                                '[' => loop_count += 1,
                                ']' => loop_count -= 1,
                                _ => {}
                            }
                        }
                    }
                }
                ']' => {
                    if self.memory[self.ptr] != 0 {
                        let mut loop_count = 1;
                        while loop_count > 0 {
                            self.ip -= 1;
                            match self.source[self.ip] as char {
                                '[' => loop_count -= 1,
                                ']' => loop_count += 1,
                                _ => {}
                            }
                        }
                        self.ip -= 1;
                    }
                }
                _ => {}
            }
            self.ip += 1;
        }
    }
}

fn main() -> io::Result<()> {
    let args: Vec<String> = env::args().collect();
    let mut interpreter = Interpreter::new();

    if args.len() == 1 {
        println!("\n **** Adderall: A Brainfuck Interpreter ****\n");
        println!("  1. Type or paste in brainfuck source code");
        println!("  2. Use Ctrl-D (Unix) or Ctrl-Z (Windows) to run the code");
        println!("  3. Use rustc adderall.rs && ./adderall [filename] to execute source file");

        io::stdin().read_to_end(&mut interpreter.source)?;
        interpreter.execute();
        println!();
    } else if args.len() == 2 {
        let mut file = File::open(&args[1])?;
        file.read_to_end(&mut interpreter.source)?;
        interpreter.execute();
    }
    Ok(())
}
